tools/pylib/boututils/datafile.py

```diff
@@ -571,49 +597,100 @@ def keys(self):
     def dimensions(self, varname):
         """Array of dimension names"""
         var = self.handle[varname]
-        vartype = str(var.attrs['type'],encoding='utf-8')
+        vartype = str(var.attrs['type'], encoding='utf-8')
         if vartype == 'Field3D_t':
-            return ('t','x','y','z')
+            return ('t', 'x', 'y', 'z')
         elif vartype == 'Field2D_t':
-            return ('t','x','y')
+            return ('t', 'x', 'y')
         elif vartype == 'scalar_t':
             return ('t')
         elif vartype == 'Field3D':
-            return ('x','y','z')
+            return ('x', 'y', 'z')
         elif vartype == 'Field2D':
-            return ('x','y')
+            return ('x', 'y')
         elif vartype == 'scalar':
             return ()
         else:
             raise ValueError("Variable type not recognized")
 
+    def vartype_from_array(self, data):
+        try:
+            ndim = len(data.shape)
+        except AttributeError:
+            ndim = 0
+        if ndim == 4:
+            return 'Field3D_t'
+        elif ndim == 3:
+            # not ideal, 3d field might be time-evolving 2d field,
+            # 'Field2D_t', but can't think of a good way to distinguish
+            return 'Field3D'
+        elif ndim == 2:
+            return 'Field2D'
+        elif ndim == 1:
+            return 'scalar_t'
+        elif ndim == 0:
+            return 'scalar'
+        else:
+            raise ValueError("Unrecognized variable type, ndims=" + str(ndim))
+
     def ndims(self, varname):
         """Number of dimensions for a variable."""
-        if self.handle is None: return None
+        if self.handle is None:
+            return None
         try:
             var = self.handle[varname]
         except KeyError:
             raise ValueError("Variable not found")
-        return len(var.shape)
+        if var.size == 1:
+            # variable is a scalar, but h5py always (?) returns numpy.ndarray,
+            # so var.shape=(1,)
+            return 0
+        else:
+            return len(var.shape)
 
     def size(self, varname):
         """List of dimension sizes for a variable."""
-        if self.handle is None: return None
+        if self.handle is None:
+            return None
         try:
             var = self.handle[varname]
         except KeyError:
             return None
         return var.shape
 
-    def write(self, name, data):
+    def write(self, name, data, vartype=None):
```

Comment:
This has a different signature than the method it's overriding. Please could you bring check that this, the base class and the other derived class (`DataFile_netCDF`), all have the same function signature for this method?